server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java

```diff
@@ -373,15 +369,9 @@ private HitContext prepareNestedHitContext(SearchContext context,
             for (SearchHit.NestedIdentity nested = nestedIdentity; nested != null; nested = nested.getChild()) {
                 String nestedPath = nested.getField().string();
                 current.put(nestedPath, new HashMap<>());
-                List<?> nestedParsedSource = XContentMapValues.extractNestedValue(nestedPath, rootSourceAsMap);
-                if ((nestedParsedSource.get(0) instanceof Map) == false && hasNonNestedParent.test(nestedPath)) {
-                    // When one of the parent objects are not nested then XContentMapValues.extractValue(...) extracts the values
-                    // from two or more layers resulting in a list of list being returned. This is because nestedPath
-                    // encapsulates two or more object layers in the _source.
-                    //
-                    // This is why only the first element of nestedParsedSource needs to be checked.
-                    throw new IllegalArgumentException("Cannot execute inner hits. One or more parent object fields of nested field [" +
-                        nestedPath + "] are not nested. All parent fields need to be nested fields too");
+                List<Map<?, ?>> nestedParsedSource = XContentMapValues.extractNestedSource(nestedPath, rootSourceAsMap);
+                if (nestedParsedSource == null) {
+                    throw new IllegalStateException("Couldn't find nested source for path " + nestedPath);
```

Comment:
It's hard to provoke because it involves explicitly breaking stuff at index time - ideally this would be an assertion but it's safer as an exception as there *may* be bugs elsewhere in the code.  But if I could write a test that produced this behaviour then I'd want to fix those bugs rather than test this!